osquery/tables/applications/posix/docker.cpp

```diff
@@ -0,0 +1,866 @@
+ * one container with "id" that starts with "12345678":
+ *   SELECT * FROM docker_containers WHERE id = '1234567890abcdef'
+ *   SELECT * FROM docker_containers WHERE id = '12345678'
+ *
+ * @param tree Property tree response from docker.
+ * @param set Set that might contain prefix values.
+ * @param key Key to look for in the property tree.
+ */
+std::string getValue(const pt::ptree& tree,
+                      const std::set<std::string>& set,
+                      const std::string& key) {
+  const std::string value = tree.get<std::string>(key, "");
+  if (set.empty()) {
+    return value; // Return value from tree, if set is empty
+  }
+
+  for (const auto& entry : set) {
+    if (value.find(entry) == 0) {
+      return entry; // If entry from set is prefix of value from tree, return
+    }
+  }
+
+  return value;
+}
+
+/**
+ * @brief Utility method to retrieve labels for docker objects.
+ *
+ * @param context Query context.
+ * @param type Docker object type (container, volume etc).
+ * @param key Primary key for docker object (id, name).
+ * @param prefix URI prefix.
+ * @param suffix URI suffix.
+ * @param path Path in the response property tree where label information is
+ *        available.
+ */
+QueryData getLabels(QueryContext& context,
+                     const std::string& type,
+                     const std::string& key,
+                     const std::string& prefix,
+                     const std::string& suffix,
+                     const std::string& path) {
+  QueryData results;
+  auto keys = context.constraints[key].getAll(EQUALS);
+  for (const auto& item : keys) {
+    pt::ptree tree;
+    Status s = dockerApi(std::string(prefix + item + suffix), tree);
+    if (!s.ok()) {
+      VLOG(1) << "Error getting docker " << type << " " << item << ": "
+              << s.what();
+      continue;
+    }
+
+    try {
+      for (const auto& label : tree.get_child(path)) {
+        Row r;
+        r[key] = item;
+        r["key"] = label.first.data();
+        r["value"] = label.second.data();
+        results.push_back(r);
+      }
+    } catch (const pt::ptree_error& e) {
+      VLOG(1) << "Error getting docker " << type << " labels " << item << ": "
+              << e.what();
+    }
+  }
+
+  return results;
+}
+
+/**
+ * @brief Entry point for docker_containers table.
+ */
+QueryData genContainers(QueryContext& context) {
+  std::string query;
+  std::set<std::string> ids;
+  getQuery(context, std::string("id"), query, ids, true);
+
+  QueryData results;
+  pt::ptree tree;
+  Status s = dockerApi(std::string("/containers/json") + query, tree);
+  if (!s.ok()) {
+    VLOG(1) << "Error getting docker containers: " << s.what();
+    return results;
+  }
+
+  for (const auto& entry : tree) {
+    try {
+      const pt::ptree node = entry.second;
+      Row r;
+      r["id"] = getValue(node, ids, "Id");
+      for (const auto& name : node.get_child("Names")) {
```

Comment:
Am assuming you want me to check for count("Names") > 0 before iterating over the names?